src/client/options/mod.rs

```diff
@@ -675,43 +675,181 @@ impl Serialize for ClientOptions {
     }
 }
 
+/// Contains the options that can be set via a MongoDB connection string.
+///
+/// The format of a MongoDB connection string is described [here](https://docs.mongodb.com/manual/reference/connection-string/#connection-string-formats).
 #[derive(Debug, Default, PartialEq)]
-struct ClientOptionsParser {
+#[non_exhaustive]
+pub struct ConnectionString {
+    /// The initial list of seeds that the Client should connect to.
+    ///
+    /// Note that by default, the driver will autodiscover other nodes in the cluster. To connect
+    /// directly to a single server (rather than autodiscovering the rest of the cluster), set the
+    /// `direct_connection` field to `true`.
     pub hosts: Vec<ServerAddress>,
-    pub srv: bool,
+
+    /// The application name that the Client will send to the server as part of the handshake. This
+    /// can be used in combination with the server logs to determine which Client is connected to a
+    /// server.
     pub app_name: Option<String>,
+
+    /// The TLS configuration for the Client to use in its connections with the server.
+    ///
+    /// By default, TLS is disabled.
     pub tls: Option<Tls>,
+
+    /// The amount of time each monitoring thread should wait between performing server checks.
+    ///
+    /// The default value is 10 seconds.
     pub heartbeat_freq: Option<Duration>,
+
+    /// When running a read operation with a ReadPreference that allows selecting secondaries,
+    /// `local_threshold` is used to determine how much longer the average round trip time between
+    /// the driver and server is allowed compared to the least round trip time of all the suitable
+    /// servers. For example, if the average round trip times of the suitable servers are 5 ms, 10
+    /// ms, and 15 ms, and the local threshold is 8 ms, then the first two servers are within the
+    /// latency window and could be chosen for the operation, but the last one is not.
+    ///
+    /// A value of zero indicates that there is no latency window, so only the server with the
+    /// lowest average round trip time is eligible.
+    ///
+    /// The default value is 15 ms.
     pub local_threshold: Option<Duration>,
+
+    /// Specifies the default read concern for operations performed on the Client. See the
+    /// ReadConcern type documentation for more details.
     pub read_concern: Option<ReadConcern>,
+
+    /// The default selection criteria for operations performed on the Client. See the
+    /// SelectionCriteria type documentation for more details.
     pub selection_criteria: Option<SelectionCriteria>,
```

Comment:
As long as the transformations to create them from the raw options are lossless parsing and aggregation, I'd be inclined to go with the higher-level structs and remove or hide the components (e.g. `read_preference_tags`).  That feels more idiomatic for Rust - see for example clap, which allows mapping fairly [complex nested structures](https://github.com/clap-rs/clap/blob/v3.1.14/examples/git-derive.rs) to the cli args.

I'd be fine going the other way if consensus disagrees with me, though :)  @isabelatkinson @kmahar thoughts?

Reply:
I think I'm slightly partial to the consolidated types because they better express the relationship between the individual options. the fact that connection strings have e.g. 3 top-level options around read preference feels mainly like a shortcoming of what can be expressed with a string. I think it could be sufficient to note in the docstring for this property which connection string options can be used together to specify the read preference. I don't feel that strongly either way though.

Reply:
Yeah, I would prefer the higher level types for the reasons Kaitlin and Abraham outlined.

Reply:
Consolidated types SGTM then. Since we're going that route, should make sure to remove the fields duplicated by them.

Reply:
Updated; it looks like only `ReadPreference` and `Credential` needed consolidation, the others already did it inline.